sympy/solvers/solvers.py

```diff
@@ -1555,9 +1555,15 @@ def _expand(p):
                     u = poly.gen
                     if u != symbol:
                         try:
+                            if deg == 1:
+                                if checksol(f, symbol, soln[0]):
+                                    temp_soln = soln
                             t = Dummy('t')
                             iv = _solve(u - t, symbol, **flags)
                             soln = list(ordered({i.subs(t, s) for i in iv for s in soln}))
+                            soln[:] = (val for val in soln if val != S.NaN)
+                            if not soln:
+                                soln = temp_soln
```

Comment:
`temp_soln` contains solutions to the polynomial, not to the original equation. `soln` is meant to contain solutions to the original equation so assigning `temp_soln` to it is wrong. Also, you are still not letting `_solve()` search further after you remove an `S.NaN` (like, in the case of `x**(2/x)-3*x*(1/x)+2`: you'll miss 1 as a solution).